File: app/src/main/java/com/example/mvvm_artichecture_sample/base/network/baseobserver/ErrorConsumer.java
package com.example.mvvm_artichecture_sample.base.network.baseobserver;

import com.example.mvvm_artichecture_sample.base.network.model.APIError;
import com.example.mvvm_artichecture_sample.base.network.statuscode.ClientCode;

import java.net.SocketTimeoutException;
import java.net.UnknownHostException;

import io.reactivex.functions.Consumer;

public class ErrorConsumer implements Consumer<Throwable> {
	
	
	private WebServiceListener listener;
	private String requestType;
	
	public ErrorConsumer(WebServiceListener listener, String requestType) {
		this.listener = listener;
		this.requestType = requestType;
	}
	
	@Override
	public void accept(Throwable throwable) throws Exception {
		APIError apiError = new APIError();
		int code;
		if (throwable instanceof UnknownHostException || throwable instanceof SocketTimeoutException) {
			apiError.setCode(ClientCode.NO_INTERNET_TIMEOUT_CODE);
			code = ClientCode.NO_INTERNET_TIMEOUT_CODE;
		} else {
			apiError.setCode(ClientCode.SERVER_TIMEOUT_CODE);
			code = ClientCode.SERVER_TIMEOUT_CODE;
		}
		apiError.setMessage("Unexpected error");
		manageOnError(apiError, code, requestType);
	}
	
	private void manageOnError(APIError apiError, int statusCode, String requestType) {
		try {
			if (listener != null) {
				listener.onError(apiError, statusCode, requestType);
			} else {
				apiError = new APIError("Unexpected error", statusCode);
				listener.onError(apiError, statusCode, requestType);
			}
		} catch (Exception e) {
			manageInternalException(e);
		}
		
	}
	
	private void manageInternalException(Exception ex) {
		ex.printStackTrace();
		if (listener != null) {
			try {
				APIError apiError = new APIError("Unexpected error", ClientCode.EXCEPTION_CODE);
				listener.onError(apiError, ClientCode.EXCEPTION_CODE, requestType);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}
}

File: app/src/main/java/com/example/mvvm_artichecture_sample/base/network/baseobserver/ErrorConsumer.kt
package com.example.mvvm_artichecture_sample.base.network.baseobserver

import com.example.mvvm_artichecture_sample.base.network.model.APIError
import com.example.mvvm_artichecture_sample.base.network.statuscode.ClientCode

import java.lang.ref.WeakReference
import java.net.SocketTimeoutException
import java.net.UnknownHostException

import io.reactivex.functions.Consumer

class ErrorConsumer(listener: WebServiceListener, private val requestType: String) : Consumer<Throwable> {

    private val listenerWeakReference: WeakReference<WebServiceListener>

    init {
        listenerWeakReference = WeakReference(listener)
    }

    @Throws(Exception::class)
    override fun accept(throwable: Throwable) {
        val apiError = APIError()
        val code: Int
        if (throwable is UnknownHostException || throwable is SocketTimeoutException) {
            apiError.code = ClientCode.NO_INTERNET_TIMEOUT_CODE
            code = ClientCode.NO_INTERNET_TIMEOUT_CODE
        } else {
            apiError.code = ClientCode.SERVER_TIMEOUT_CODE
            code = ClientCode.SERVER_TIMEOUT_CODE
        }
        apiError.message = "Unexpected Error"
        manageOnError(apiError, code, requestType)
    }

    private fun manageOnError(apiError: APIError, statusCode: Int, requestType: String) {
        try {
            listenerWeakReference.get()!!.onError(apiError, statusCode, requestType)
        } catch (e: Exception) {
            manageInternalException(e)
        }

    }

    private fun manageInternalException(ex: Exception) {
        ex.printStackTrace()
        if (listenerWeakReference.get() != null) {
            try {
                val apiError = APIError("Unexpected Error", ClientCode.EXCEPTION_CODE)
                listenerWeakReference.get()!!.onError(apiError, ClientCode.EXCEPTION_CODE, requestType)
            } catch (e: Exception) {
                e.printStackTrace()
            }

        }
    }
}
